Add precommit check for the layout pattern for security auditing

albertzaharovits opened the pull request:
`x-pack:plugin:core` contains a `log4j2.properties` conf file that must be used for the ES Security audit logs (the audit log uses the log4j2 library and also requires a special layout pattern, which is defined in the conf file). We have unit tests that validate this layout pattern, which is used in the archive builds.

The problem is that the `log4j2.properties` conf file has to be slightly different for docker builds. Probably the most important reason for that is that in a docker environment the logs are all outputted to the console. But the layout pattern should not be different in the two situations. Whenever the layout pattern changes to accommodate new audit record fields, the changes must be reflected to both files.

This PR adds the audit record fields from #58928 to the docker build's layout pattern config. This was a mistake, those changes should've been replicated back then.
It also adds a verification task to the `distribution:docker` project which would catch future such problems.

albertzaharovits committed:
Done

elasticmachine commented:
Pinging @elastic/es-security (:Security/Audit)

mark-vieira commented on distribution/docker/build.gradle:
```diff
@@ -165,6 +168,30 @@ tasks.register("copyKeystore", Sync) {
   }
 }
 
+tasks.register("ensureLog4jConfigurationSync") {
```
I wonder if we should wire this to `precommit` so it fails early. Otherwise this will only fail when we actually go to build these docker images.

mark-vieira commented on distribution/docker/build.gradle:
```diff
@@ -165,6 +168,30 @@ tasks.register("copyKeystore", Sync) {
+    def originalLog4j = project(":x-pack:plugin:core").file('src/main/config/log4j2.properties').getBytes()
+    def checksumOriginalLog4j = Base64.getEncoder().encodeToString(MessageDigest.getInstance("MD5").digest(originalLog4j))
+    if (false == validatedChecksumOriginalLog4j.equals(checksumOriginalLog4j)) {
+      throw new GradleException("The src/main/config/log4j2.properties file from the x-pack:plugin:core has been changed. " +
```
We should give some indication as to where "this task's declaration" lives. Perhaps provide the relative path to this build.gradle file in the exception message? Or instead externalize the checksum in a file (like we do for dependency artifacts) and just point to that so folks don't actually have to change build scripts.

mark-vieira reviewed (commented):
Is there any way to be a bit more intelligent here? Is the docker file expected to be as strict subset, or superset? Do we know the lines that will be different and perhaps can just compare the rest?

albertzaharovits commented:
Thanks for the fast turn-around @mark-vieira !

> Is there any way to be a bit more intelligent here? Is the docker file expected to be as strict subset, or superset? Do we know the lines that will be different and perhaps can just compare the rest?

I dug around today but I was unable to find a reasonable solution **in the log4j2 realm.** I'll explain:

Essentially, the `log4j2.properties` from `distribution/docker` is similar to what is stitched together in the archive builds (by concatenating fragments from modules), but the appender types are all `Console` instead of `RollingFile`.

In general, the changes to these two `log4j2.properties` are infrequent and might not be replicated between the two (eg adding a parameter to the `RollingFileAppender`). But changes to the audit logger appender's `pattern` must be reflected every time.

Now, pulling the `pattern` configuration out separately, so it's only one place to change, (eg on it's own file) turns out is quite a challenge.
For a start, to break apart the existing config into several ones, we need to either set a [java system property](https://logging.apache.org/log4j/2.x/manual/configuration.html#CompositeConfiguration) or rewrite the configs in [XML](https://logging.apache.org/log4j/2.x/manual/configuration.html#XInclude). But because appenders cannot be configured without a layout (and that the layout cannot be set or changed at runtime), we'll also need to use a separate properties file and have the appenders refer to the external property value; this is overkill IMO and clearly takes away the expressivity of a separate log4j2.properties file.

Another option that I've explored is to create log4j2 plugins that define the layout (the layout is still a text resource file but it is exposed by the plugin). We have examples of plugins in the current codebase, but they need to sit in the `server` project or as a jar library. Hence the plugin route is again impractical.

I feel that any solution is the log4j2 realm would be unnecessary complicated.

Can we instead recur to some gradle magic?
**What would you say of a precommit check in the `distribution/docker` project that parses the configs as property files (which they are) and (only) ensures that the two homologous patterns are exactly the same between the files?**

mark-vieira commented:
I'm going to go ahead and ping @breskeby here just to have more brains looking at this problem. Perhaps there are trivial solution I am overlooking.

breskeby commented:
Couldn't we just read the `appender.audit_rolling.layout.pattern` property from both files as properties and check they match?

albertzaharovits commented:
> Couldn't we just read the appender.audit_rolling.layout.pattern property from both files as properties and check they match? Should that property not be exactly the same across both configs?

Yes, that should do it, that's what I was thinking as well; read the two files as property files and check for string match of these two specific property values (that are named the same in the two log4j2 configs).
I think hooking the check into the procommit phase, as Mark was suggesting, is more desirable and appropriate than the build phase, no? It's OK if the test sits in the distribution/docker project's build file, i.e. where the log4j2 config is replicated to, rather than the build of the `x-pack:plugin:core` project, where the other log4j2 file resides, which is the natural place where modifications to the pattern go?

breskeby commented:
what mark means by precommit phase is that the `precommit` task should depend on this check. I'm fine with having that check being located in `distributions/docker/build.gradle`

albertzaharovits commented:
Thank you @breskeby ! I'll update this PR accordingly.

albertzaharovits committed:
Merge branch 'master' into audit_log4j_configuration_sync_up

albertzaharovits committed:
check by property name

albertzaharovits committed:
Revert prev

albertzaharovits commented:
@breskeby I've updated this PR with the precommit check for the equality of the two property values. I've also moved the log4j2.properties changes over to a different PR https://github.com/elastic/elasticsearch/pull/63516 , hence this PR will fail until that gets merged.

This is ready for review.

mark-vieira commented on distribution/docker/build.gradle:
```diff
@@ -165,6 +165,41 @@ tasks.register("copyKeystore", Sync) {
+    }
+
+    if (false == coreLog4jProperties.containsKey(patternPropertyKey)) {
+      throw new GradleException("The [" + originalLog4j.getPath() + "] file changed such that the layout pattern is not " +
```
This is Groovy, so need to do string concatination, just use groovy string interpolation (ex: `The [${originalLog4j.path}] file...`).

albertzaharovits committed:
Merge branch 'master' into audit_log4j_configuration_sync_up

albertzaharovits committed:
Inputs to task as per Mark's review

albertzaharovits commented on distribution/docker/build.gradle:
```diff
@@ -165,6 +165,41 @@ tasks.register("copyKeystore", Sync) {
   }
 }
 
+tasks.register("checkSecurityAuditLayoutPatternIdentical") {
```
Thanks Mark.

breskeby commented on distribution/docker/build.gradle:
```diff
@@ -165,6 +165,43 @@ tasks.register("copyKeystore", Sync) {
+tasks.register("checkSecurityAuditLayoutPatternIdentical") {
+  // the two log4j2.properties files containing security audit configuration for archive and docker builds respectively
+  def originalLog4j = project(":x-pack:plugin:core").file('src/main/config/log4j2.properties')
+  def dockerLog4j = project.projectDir.toPath().resolve("src/docker/config/log4j2.properties").toFile()
```
you can just do `project.file("src/docker/config/log4j2.properties")`

albertzaharovits replied:
Thanks 🤦

breskeby commented on distribution/docker/build.gradle:
```diff
@@ -165,6 +165,43 @@ tasks.register("copyKeystore", Sync) {
+    }
+  }
+}
+precommit.dependsOn 'checkSecurityAuditLayoutPatternIdentical'
```
Please use `tasks.named("precommit").configure { it.dependsOn('checkSecurityAuditLayoutPatternIdentical') }` to make use of gradles "task avoidance api" to avoid creating tasks early. see https://docs.gradle.org/current/userguide/task_configuration_avoidance.html for further information

albertzaharovits replied:
Ha, I was wondering what's up with the `register` methods, thanks for the link!
I've changed it to:
```
tasks.named("precommit").configure {
  dependsOn 'checkSecurityAuditLayoutPatternIdentical'
}
```

breskeby reviewed (changes requested):
lgtm. just some minor change requests

albertzaharovits committed:
rene's review

albertzaharovits committed:
Merge branch 'master' into audit_log4j_configuration_sync_up

breskeby reviewed (approved):
👍

albertzaharovits commented:
Thank you Mark and Rene for the guidance and tips!

breskeby commented:
@albertzaharovits can you ensure this is back ported to the 7.x branch?

albertzaharovits commented:
@breskeby I did that already by cherry-picking and pushing to origin on the 7.x branch(https://github.com/elastic/elasticsearch/commit/a5c96f3c7b42b9c1ecdd9d3be97b9fc1d656c300) . I've tested locally the precommit task. I thought a backport PR was unnecessary.

breskeby commented:
👍
